### ISSUE TITLE
`num_tokens` overcounted under TP

### ISSUE THREAD

**Issue by qgallouedec** [huggingface/trl#7100]
Under tensor parallelism every `tp` rank receives the same batch, but the `num_tokens` metric is summed over all processes, so each token is counted `tp_size` times.

```python
# torchrun --nproc_per_node 2 repro.py
from datasets import Dataset
from transformers import AutoModelForCausalLM
from transformers.distributed import DistributedConfig

from trl import SFTConfig, SFTTrainer

model = AutoModelForCausalLM.from_pretrained(
    "trl-internal-testing/tiny-Qwen3ForCausalLM",
    distributed_config=DistributedConfig(tp_size=2),
)
dataset = Dataset.from_dict({"prompt": ["hello world"] * 8, "completion": [" foo"] * 8})
trainer = SFTTrainer(
    model=model,
    args=SFTConfig(max_steps=1, per_device_train_batch_size=1, logging_steps=1, report_to=[]),
    train_dataset=dataset,
)
trainer.train()
```

### Outcome

`num_tokens` is twice the number of tokens actually trained on: the two ranks hold the same sample, and the metric adds both copies. With `tp_size=32` it is 32x, which makes the logged tokens/s meaningless for anyone using tensor or expert parallelism.

### Expected behavior

`num_tokens` counts each token once, dividing by the tensor-parallel size, as `Trainer` already does for `num_input_tokens_seen` and for `num_items_in_batch`.

### System info

- Platform: Linux-6.8.0-1057-aws-x86_64-with-glibc2.35
- Python version: 3.13.13
- TRL version: 1.10.0.dev0+2b9669e
- PyTorch version: 2.13.0
- accelerator(s): NVIDIA H100 80GB HBM3
- Transformers version: 5.16.0.dev0
- Accelerate version: 1.14.0
- Accelerate config: not found
- Datasets version: 5.0.0
- HF Hub version: 1.29.0
- bitsandbytes version: 0.50.1
- DeepSpeed version: 0.19.3
- Liger-Kernel version: not installed
- PEFT version: 0.19.1
- vLLM version: 0.28.0
